sqlccl, distsqlrun: use unbuffered distsql router during IMPORT

madelynnblue opened the pull request:
Previously all distsql routers would buffer to prevent some deadlock
scenarios. However during IMPORT, data is produced faster than it is
consumed, and this buffer would result in SQL terminating the query
due to memory constraints.

Add a field in routers to disable buffering. Implement this by
refactoring common routerBase code out of the routers (otherwise
there's a lot of duplicate variable passing around).

Fixes #20761

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20901)
<!-- Reviewable:end -->

madelynnblue committed:
sqlccl, distsqlrun: use unbuffered distsql router during IMPORT

Previously all distsql routers would buffer to prevent some deadlock
scenarios. However during IMPORT, data is produced faster than it is
consumed, and this buffer would result in SQL terminating the query
due to memory constraints.

Add a field in routers to disable buffering. Implement this by
refactoring common routerBase code out of the routers (otherwise
there's a lot of duplicate variable passing around).

Fixes #20761

Release note: None

RaduBerinde commented:
LGTM

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/routers.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20901#-L0kn-HdC3eSLLCsbw5J:-L0kn-HdC3eSLLCsbw5K:b-gbof4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/dea58d6dc46de1298e896d963acbaa8e3760d39c/pkg/sql/distsqlrun/routers.go#L185)):*
> ```Go
> 	n := len(streams)
> 	if disableBuffering {
> 		n = 1
> ```

This needs a comment ("by starting the semaphore at 1, the producer is blocked whenever one of the streams are blocked"), and a `TODO: instead of disabling buffering this way, we should short-circuit the entire router implementation and push directly to the output stream`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20901#-:-L0knpo7OqvU3IYOpqvz:bb6ywmx)*
<!-- Sent from Reviewable.io -->
